purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt

```diff
@@ -1241,23 +1249,42 @@ class Purchases internal constructor(
             )
         }
 
-    private fun getProducts(
+    private fun getProductsOfTypes(
+        productIds: Set<String>,
+        types: Set<ProductType>,
+        callback: GetStoreProductsCallback
+    ) = getProductsOfTypes(productIds, types, emptyList(), callback)
+
+    private fun getProductsOfTypes(
         productIds: Set<String>,
-        productType: ProductType,
+        types: Set<ProductType>,
+        collectedStoreProducts: List<StoreProduct>,
         callback: GetStoreProductsCallback
     ) {
-        billing.queryProductDetailsAsync(
-            productType,
-            productIds,
-            { storeProducts ->
-                dispatch {
-                    callback.onReceived(storeProducts)
-                }
-            }, {
-                dispatch {
-                    callback.onError(it)
-                }
-            })
+        val typesRemaining = types.filter { it != ProductType.UNKNOWN }.toMutableSet()
+        val type = typesRemaining.firstOrNull()?.also { typesRemaining.remove(it) }
```

Comment:
I find this code very complex and hard to understand (maybe it's just that I'm not used to the Kotlin common idioms)
My understanding is it's basically looping over the typesRemaining list through the recursive call in L1273 and by popping an element from the list.
The fact that there's no explicit stop condition (like I would expect in a recursive call) makes it even harder for me. I understand that `type?.let` is doing that, but it's hard to see at first glance.
The removal of UNKNOWN is also done in all calls, instead of just in the first one, which is a bit confusing.

Isn't there a way to just write this as a flatMap over types?

Reply:
okay so i kinda see stefan's point. as much as i liked the recursion maybe it's not readable/idiomatic

before you had the user pass a set of types too right? maybe the recursion felt more natural when it was more than one...

though i think the change to one makes sense, as there's only 2 options and thye can just omit the parameter for both

Reply:
This is definitely _not_ the most ideal approach to this but its is weirdly less complex and more efficient than what it was doing.

Previously we had a callback in a callback to fetch all SUBS and then INAPPs and then filtering on that. This approach allows for filtering both _or_ just one... or more if that ever happens 🤷‍♂️ 

But we can't really just map over these because they calls are async (with the callback) which is why we either needed to:
1) Hardcoded callback in a callback like before this PR
2) Recursion what this PR does

There is an option to do something with coroutines but we haven't done that yet in the SDK so it isn't really the best option for this PR 🤷‍♂️ 

I wouldn't say that recursion is the "best" pattern but it is something that I do run into occasionally 😇 

>The fact that there's no explicit stop condition 

I can make a more explicit stop for the recurrision but i think the `type?.let` is the most Kotlin way of doing that.

But... I really wish we could just do like an async/await and flatMap but that is not in the cards 😔

Reply:
I think this is an alternative, not necessarily better, but just wanted to throw it out there:

```kotlin
    private fun getProductsOfTypes(
        productIds: Set<String>,
        types: Set<ProductType>,
        callback: GetStoreProductsCallback
    ) {
        val collectedStoreProducts = listOf<StoreProduct>()
        var typesToReturn = types.toSet()
        types.toMutableSet().forEach {
            billing.queryProductDetailsAsync(
                it,
                productIds,
                { storeProducts ->
                    collectedStoreProducts + storeProducts
                    typesToReturn = typesToReturn.toMutableSet() - it
                    if (typesToReturn.isEmpty()) {
                        dispatch {
                            callback.onReceived(collectedStoreProducts)
                         }
                    }
                }, {
                    dispatch {
                        callback.onError(it)
                    }
                })
        }
    }
 ```
 
It might not be thread safe, I am not sure, but I think it should work

Reply:
I wasn't sure how billing client handle multiple query calls at once which is why I went with recursion to not worry about any threading issues 🤷‍♂️ This starting to feel like an interviewing problem 🙈